### ISSUE TITLE
Socket Error: Did not receive #handshake from client before timeout

### ISSUE THREAD

**Issue by McFarts** [SocketCluster/socketcluster#107]
I keep getting:
```
 [Notice] Error: Socket Error: Did not receive #handshake from client before
timeout
```

My worker.js:
```
module.exports.run = function (worker) {
  console.log('   >> Worker PID:', process.pid);
  
  //var app = require('express')();
  
  //var httpServer = worker.httpServer;
  var scServer = worker.scServer;
  
 // app.use(serveStatic(path.resolve(__dirname, 'public')));

  //httpServer.on('request', app);

  var count = 0;

  /*
    In here we handle our incoming realtime connections and listen for events.
  */
  scServer.on('connection', function (socket) {
  
	//console.log(socket);
	
    // Some sample logic to show how to handle client events,
    // replace this with your own logic
   socket.emit('ping', {message: 'This is an object with a message property'});
	
	
	/*
    socket.on('sampleClientEvent', function (data) {
      count++;
      console.log('Handled sampleClientEvent', data);
      scServer.global.publish('sample', count);
    });
    

    var interval = setInterval(function () {
      socket.emit('rand', {
        rand: Math.floor(Math.random() * 5)
      });
    }, 1000);
    
    socket.on('disconnect', function () {
      clearInterval(interval);
    });
	
	*/
	
	socket.send('hello');
	
	
	
  });
};
```

and server.js:
```
var SocketCluster = require('socketcluster').SocketCluster;
var numCPUs = require('os').cpus().length;



var socketCluster = new SocketCluster({
  workers: numCPUs, // Number of worker processes
  brokers: 1, // Number of broker processes
  port: 9300, // The port number on which your server should listen
  appName: 'myapp', // A unique name for your app

  /* A JS file which you can use to configure each of your 
   * workers/servers - This is where most of your backend code should go
   */
  workerController: __dirname + '/worker.js',

  /* JS file which you can use to configure each of your 
   * brokers - Useful for scaling horizontally across multiple machines (optional)
   */ 
  // brokerController: __dirname + '/broker.js',
  
  
  
  
  

  // Whether or not to reboot the worker in case it crashes (defaults to true)
  rebootWorkerOnCrash: true
});
```

I don't use express as this is only a gameserver.  Any idea? :)

**Comment by jondubois** [comment]
@McFarts You should check the client and server versions to make sure that they are compatible.
The latest server is 2.3.14 and the client is 2.3.13.

To find out your current client version, just enter ```socketCluster.version``` in the dev console. The server version comes up in the console when you launch the SC server.

**Comment by ccravens** [comment]
Is it required that we use the SocketCluster client library on the browser or can we use standard WebSocket connections?

**Comment by jondubois** [comment]
@ccravens SC adds its own protocol layer on top of the WebSocket protocol so you can't really use the plain WebSocket API to connect to it unless you're prepared to re-implement part of the SC protocol.
For example, SC adds a handshake step which it uses to do JWT token authentication and set up channel subscriptions.

In theory, you could make a more lightweight version of the SC client which only implements the basic features required but such client doesn't currently exist.

Out of interest, why can't you use the current client?

**Comment by ccravens** [comment]
Hello Jon:

The project we're working on requires an minimalist approach, due to the fact that we are implementing so many different technologies and have a requirement to make the code base as small and efficient as possible. This also includes re-implementing some of the angular core-features from scratch such as the 2-way data binding and directives. We also require an extreme fine-grain control of every process. Not to say that the SocketCluster client is inefficient in any way, however we will probably peel back the client and take a look at how it is implemented. SocketCluster, and distributed real-time communication with pub/sub channels is the backbone of what we are working on. If we do end up creating a minimalist, light-weight version of SocketClient and it is useful we would be happy to release back to the community.

Thanks for your quick response!

**Comment by tilleps** [comment]
I am able to reproduce this by disabling cookies in the browser (Chrome/Firefox/Safari).  The error throws about 8-9 seconds after what appears to be a successful handshake (able to emit/publish).

I'm guessing the lack of cookie headers is preventing **SCServer._handleHandshakeTimeout** from being cleared out?

**Comment by jondubois** [comment]
Sounds like an SC1 issue.

**Comment by tilleps** [comment]
Reproduced the problem with SocketCluster version 2.3.14 and cookies disabled.  

It seems that the disabling of cookies also disables the usage of localStorage which is causing the interruption.

On Firefox, I get this error:
`SecurityError: The operation is insecure.  socketcluster.js line 40`

```javascript
AuthEngine.prototype.loadToken = function (name, callback) {
  var token;
  
  if (global.localStorage) {
    token = global.localStorage.getItem(name);
  } else {
    token = this._internalStorage[name] || null;
  }
  callback(null, token);
};
```

**Comment by jondubois** [comment]
@tilleps Nice catch. Fixed in v2.3.17.
